File: 20150531-StringCalculator-Scala/src/test/scala/kata/SCTest.scala
package kata

import kata.SC.add
import org.scalacheck.Gen
import org.scalatest._
import org.scalatest.prop.PropertyChecks

class SCTest extends FlatSpec with PropertyChecks with Matchers {
	val integers = Gen.choose(Int.MinValue, Int.MaxValue)
	val listOfIntegers = Gen.listOf(integers)
	val separators = Gen.choose(0.toChar, 125.toChar)
	val stringSeparators = for (c <- Gen.listOf(separators)) yield c.filter(validSeparator).mkString
	val listOfStringSeparators = for (c <- Gen.listOf(stringSeparators)) yield c.filter(_.length > 0)
	val nonNegativeIntegers = Gen.choose(0, Int.MaxValue)
	val listOfNonNegativeIntegers = Gen.listOf(nonNegativeIntegers)

	"Given an empty string" should "return 0" in
		assert(add("") == 0)

	"Given a non-negative integer as a string" should "return the value if less than 1001 otherwise 0" in
		forAll (nonNegativeIntegers) { n => 
			val result = if (n < 1001) n else 0
			assert(add(n.toString) == result) 
		}

	"Given non-negative integers separated by a comma" should "return the sum of those less than 1001" in
		forAll (listOfNonNegativeIntegers) { ns =>
			assert(add(ns.mkString(",")) == sum(ns))
		}

	"Given non-negative integers separated by a comma or newline" should "return the sum of those less than 1001" in
		forAll (listOfNonNegativeIntegers) { ns =>
			assert(add(mkString(ns, List(",", "\n"))) == sum(ns))
		}

	"Given non-negative integers separated by a single character custom separator" should "return the sum of those less than 1001" in
		forAll (listOfNonNegativeIntegers, separators) { (ns, sep) =>
			whenever(ns.length > 0 && validSeparator(sep)) {
				assert(add(s"//$sep\n${ns.mkString(sep.toString)}") == sum(ns))
			}
		}

	"Given non-negative integers separated by a multiple multi-character custom separator" should "return the sum of those less than 1001" in
		forAll (listOfNonNegativeIntegers, listOfStringSeparators) { (ns, seps) =>
			whenever(ns.length > 0 && seps.length > 0) {
				assert(add(s"//[${seps.mkString("][")}]\n${mkString(ns, seps)}") == sum(ns))
			}
		}

	"Given numbers with at least one negative" should "throw an exception with the negatives numbers in the message" in
		forAll(listOfIntegers) { ns =>
			whenever (ns.exists(_ < 0)) {
				val ex = intercept[java.lang.IllegalArgumentException] {
					add(ns.mkString(","))
				}
				assert(ex.getMessage == ns.filter(_ < 0).mkString(","))
			}
		}

	def validSeparator(ch: Char): Boolean = ch != '-' && !Character.isDigit(ch) && ch != '[' && ch != ']' && ch != 10.toChar && ch != 13.toChar

	def sum(ns: List[Int]): Int = ns.filter(_ < 1001).sum

	def mkString(ns: List[Int], seps: List[String]): String = ns match {
		case Nil => ""
		case x::Nil => x.toString
		case x1::xs => x1.toString ++ Gen.oneOf(seps).sample.get ++ mkString(xs, seps)
	}
}

File: 20180614-StringCalculator-Kotlin/src/test/kotlin/kata/SCTest.kt
package kata

import io.kotlintest.properties.Gen
import io.kotlintest.properties.assertAll
import io.kotlintest.shouldBe
import io.kotlintest.shouldThrow
import io.kotlintest.specs.StringSpec


val ListOfIntegers =
        Gen.list(Gen.int())

val NonNegativeIntegers =
        Gen.nats()

val ListOfNonNegativeIntegers =
        Gen.list(NonNegativeIntegers)

val Separators =
        Gen.choose(0, 127).map { it.toChar() }.filter { !it.isDigit() && it != '[' }

val StringSeparators =
        Gen.string().filter { it.isNotBlank() && it.all { !it.isDigit() && it != '[' } }

val ListOfStringSeparators =
        Gen.list(StringSeparators).filter { it.isNotEmpty() }


class SCTest : StringSpec({
    "blank should return 0" {
        add("").shouldBe(0)
    }

    "given a non-negative integer should return its value" {
        assertAll(NonNegativeIntegers) { n ->
            add(n.toString()).shouldBe(n)
        }
    }

    "given non-negative integers separated with a comma or newline should return the sum" {
        assertAll(ListOfNonNegativeIntegers) { ns ->
            add(ns.mkString(listOf(",", "\n"))).shouldBe(ns.sum())
        }
    }

    "given non-negative integers separated with a single custom character should return the sum" {
        assertAll(ListOfNonNegativeIntegers, Separators) { ns, sep ->
            add("//$sep\n${ns.joinToString(sep.toString())}").shouldBe(ns.sum())
        }
    }

    "given non-negative integers separated with multiple multiple character should return the sum" {
        assertAll(ListOfNonNegativeIntegers, ListOfStringSeparators) { ns, seps ->
            add("//[${seps.joinToString("][")}]\n${ns.mkString(seps)}").shouldBe(ns.sum())
        }
    }

    "given integers with at least one negative should throw an exception with all of the negatives in the message" {
        assertAll(ListOfIntegers.filter { it.any { it < 0 } }) { ns ->
            val exception = shouldThrow<IllegalArgumentException> { add(ns.joinToString(",")) }

            exception.message.shouldBe(ns.filter { it < 0 }.joinToString(","))
        }
    }
})


fun List<Int>.mkString(separators: List<String>): String =
        this.mkString(Gen.from(separators))

fun List<Int>.mkString(separators: Gen<String>): String =
        when {
            this.isEmpty() ->
                ""

            this.size == 1 ->
                this[0].toString()

            else ->
                this[0].toString() + separators.random().first() + this.drop(1).mkString(separators)
        }
